models/mgan.py

```diff
@@ -0,0 +1,83 @@
+                weight[i].append(1 - relative_pos / sentence_len)
+            for j in range(pos_inx[i][0], pos_inx[i][1] + 1):
+                weight[i].append(0)
+            for j in range(pos_inx[i][1] + 1, seq_len):
+                relative_pos = j - pos_inx[i][1]
+                aspect_len = pos_inx[i][1] - pos_inx[i][0] + 1
+                sentence_len = seq_len - aspect_len
+                weight[i].append(1 - relative_pos / sentence_len)
+        weight = torch.tensor(weight)
+        return weight
+
+class MGAN(nn.Module):
+    def __init__(self, embedding_matrix, opt):
+        super(MGAN, self).__init__()
+        self.opt = opt
+        self.embed = nn.Embedding.from_pretrained(torch.tensor(embedding_matrix, dtype=torch.float))
+        self.ctx_lstm = DynamicLSTM(opt.embed_dim, opt.hidden_dim, num_layers=1, batch_first=True, bidirectional=True)
+        self.location = LocationEncoding(opt)
+        self.asp_lstm = DynamicLSTM(opt.embed_dim, opt.hidden_dim, num_layers=1, batch_first=True, bidirectional=True)
+        self.dense = nn.Linear(8 * opt.hidden_dim, opt.polarities_dim)
+
+    def forward(self, inputs):
+        text_raw_indices = inputs[0] # batch_size x seq_len
+        aspect_indices = inputs[1] 
+        text_left_indices= inputs[2]
+        ctx_len = torch.sum(text_raw_indices != 0, dim=1)
+        asp_len = torch.sum(aspect_indices != 0, dim=1)
+        left_len = torch.sum(text_left_indices != 0, dim=-1)
+        aspect_in_text = torch.cat([left_len.unsqueeze(-1), (left_len+asp_len-1).unsqueeze(-1)], dim=-1)
+
+        ctx = self.embed(text_raw_indices) # batch_size x seq_len x embed_dim
+        asp = self.embed(aspect_indices) # batch_size x seq_len x embed_dim
+
+        ctx_out, (_, _) = self.ctx_lstm(ctx, ctx_len) 
+        ctx_out = self.location(ctx_out, aspect_in_text) # batch_size x (ctx)seq_len x 2*hidden_dim
+        ctx_pool = torch.sum(ctx_out, dim=1)
+        ctx_pool = torch.div(ctx_pool, ctx_len.float().unsqueeze(-1)).unsqueeze(-1) # batch_size x 2*hidden_dim x 1
+
+        asp_out, (_, _) = self.asp_lstm(asp, asp_len) # batch_size x (asp)seq_len x 2*hidden_dim
+        asp_pool = torch.sum(asp_out, dim=1)
+        asp_pool = torch.div(asp_pool, asp_len.float().unsqueeze(-1)).unsqueeze(-1) # batch_size x 2*hidden_dim x 1
+
+        alignment_mat = torch.matmul(ctx_out, asp_out.transpose(1, 2)) # batch_size x (ctx)seq_len x (asp)seq_len
+        # batch_size x 2*hidden_dim
+        f_asp2ctx = torch.matmul(ctx_out.transpose(1, 2), F.softmax(alignment_mat.max(2, keepdim=True)[0], dim=1)).squeeze(-1)
+        f_ctx2asp = torch.matmul(F.softmax(alignment_mat.max(1, keepdim=True)[0], dim=2), asp_out).transpose(1, 2).squeeze(-1) 
+
+        c_asp2ctx = torch.matmul(ctx_out.transpose(1, 2), F.softmax(torch.matmul(ctx_out, asp_pool), dim=1)).squeeze(-1) 
+        c_ctx2asp = torch.matmul(asp_out.transpose(1, 2), F.softmax(torch.matmul(asp_out, ctx_pool), dim=1)).squeeze(-1)
```

Comment:
Compared with the original paper, attention weight matrix W_{ca} and W_{cc} is missing?

Reply:
Yes, the original paper calculate attention weight in bi-linear way while I'm using dot product.
Would it be better if I fix them?

Reply:
Yes, the paper calculate attention weight in bi-linear way while I'm using dot product.
Would it be better if I fix them?